modules/core/04-channel/v2/keeper/msg_server_test.go

```diff
@@ -433,3 +434,97 @@ func (suite *KeeperTestSuite) TestMsgAcknowledgement() {
 		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestMsgTimeout() {
+	var (
+		path   *ibctesting.Path
+		packet channeltypesv2.Packet
+	)
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			name:     "success",
+			malleate: func() {},
+		},
+		{
+			name: "failure: no-op",
+			malleate: func() {
+				suite.chainA.App.GetIBCKeeper().ChannelKeeperV2.DeletePacketCommitment(suite.chainA.GetContext(), packet.SourceChannel, packet.Sequence)
+
+				// Modify the callback to return a different error.
+				// This way, we can verify that the callback is not executed in a No-op case.
+				path.EndpointA.Chain.GetSimApp().MockModuleV2A.IBCApp.OnTimeoutPacket = func(context.Context, string, string, channeltypesv2.PacketData, sdk.AccAddress) error {
+					return mock.MockApplicationCallbackError
+				}
+			},
+			expError: channeltypesv1.ErrNoOpMsg,
+		},
+		{
+			name: "failure: callback fails",
+			malleate: func() {
+				path.EndpointA.Chain.GetSimApp().MockModuleV2A.IBCApp.OnTimeoutPacket = func(context.Context, string, string, channeltypesv2.PacketData, sdk.AccAddress) error {
+					return mock.MockApplicationCallbackError
+				}
+			},
+			expError: mock.MockApplicationCallbackError,
+		},
+		{
+			name: "failure: channel not found",
+			malleate: func() {
+				// change the source id to a non-existent channel.
+				packet.SourceChannel = "not-existent-channel"
+			},
+			expError: channeltypesv2.ErrChannelNotFound,
+		},
+		{
+			name: "failure: invalid commitment",
+			malleate: func() {
+				suite.chainA.App.GetIBCKeeper().ChannelKeeperV2.SetPacketCommitment(suite.chainA.GetContext(), packet.SourceChannel, packet.Sequence, []byte("foo"))
+			},
+			expError: channeltypesv2.ErrInvalidPacket,
+		},
+		{
+			name: "failure: unable to timeout if packet has been received",
+			malleate: func() {
+				err := path.EndpointB.MsgRecvPacket(packet)
+				suite.Require().NoError(err)
+			},
+			expError: commitmenttypes.ErrInvalidProof,
```

Comment:
added this test to ensure timing out is not possible if the packet has been received